### DISCUSSION TITLE
Current cron implementation does not respect the de facto standard

### DISCUSSION THREAD

**Discussion by mlusetti** [pocketbase/pocketbase#4394] [Q&A] [answered]
Current implementation of scheduling using cron expression fire the configured function based on the second the server has been fired.

It easy to spot, just put the job scheduling from the documentation: https://pocketbase.io/docs/js-jobs-scheduling/ in a fresh installation of pocketbase and you will se the log is fired at the second you ran the server.
Instead it should fire at second zero (0).

For example if you configure as per documentation: "*/2 * * * *" your function should be fired at minutes: 0, 2, 4, 6, 8, 10... and so on till 58 each each of which at second: 0. Right now the second at which the function is fired depends on the second you ran the "pocketbase serve"

Does anyone else noticed this?
Is there interest in a fix or the current implementation is good enough for the most ?

**Comment by ganigeorgiev** [comment]
What do you mean by "does not respect the de facto standard"?

Why not always firing on the 0 second is a problem in the first place?

The cron expression has only minute granularity. There is no hard specification (or at least I'm not aware of) where it is required to always fire on the zero second.

Also note that the cron timer is dependent on the Go `time.Ticker` and I'm not sure if we can have a reliable "always 0 second" tick as it will depend on the tick channel receiver and whether the program is busy with something else (based on the documentation the ticker will auto adjust its interval or drop ticks to compensate).

**Reply by mlusetti** [reply]
As I said is a de facto standard as any other known implementation does fire at zero second.
Every unix cron implementation follow this rule and feel unnatural to fire at the second which the server is started, this is not what everyone should expect in respect of a cron expression which tells "every two minutes".

I know the current implementation is base on timer.Ticker as I briefly looked at the code and that indeed could be a problem and indeed this one of the reason I filed a discussion and not an issue in the first place.

BTW I've solved my own problem in another way and I'm happy to leave things as they are now. I would like just to point to your attention this, somehow, weird behaviour.

**Reply by mlusetti** [reply]
If you feel the discussion title is too "strong" I'm happy to change it, let me know.

**Reply by ganigeorgiev** [reply]
I'll experiment later this week to try to adjust or delay the first tick to start from the closest `minute:00` time but I don't think it can be relied on and again the standard cron format doesn't specify seconds/milliseconds/etc. and minute is the smallest unit.

**Reply by ganigeorgiev** [reply]
Note that even with the standard OS crontab utility there is no guarantee that it will fire on the 00 second.

For the test I've added `date >> /path/to/file` to execute every minute and this is are the results:

```
21.02.2024 (ср) 18:12:01 EET
21.02.2024 (ср) 18:13:02 EET
21.02.2024 (ср) 18:14:01 EET
21.02.2024 (ср) 18:15:01 EET
21.02.2024 (ср) 18:16:02 EET
21.02.2024 (ср) 18:17:01 EET
```

_(Obviously the time will be affected by the `date` command itself)_.

As mentioned earlier I'll consider trying to make it closer to 00 second, but the point is that you can't rely on `time.Now()` or similar to report 00 second in your cron function if your use case requires it for some reason.

**Reply by ganigeorgiev** [reply, marked as answer]
I've added an extra delay timer in the [develop branch](https://github.com/pocketbase/pocketbase/commit/5a715cc60ad66ab0d9e7737631159ace79615228) to attempt starting the ticker at the 00 second. Based on my local tests it seems consistent but again there is no guarantee that it will always fire on 00 second and you may find better initializing a custom more granular timer.

```js
cronAdd("test", "* * * * *", () => {
    console.log(Date.now())
})

2024/02/21 19:43:53 Server started at http://127.0.0.1:8090
├─ REST API: http://127.0.0.1:8090/api/
└─ Admin UI: http://127.0.0.1:8090/_/

2024/02/21 19:44:00 1708537440000
2024/02/21 19:45:00 1708537500000
2024/02/21 19:46:00 1708537560001
2024/02/21 19:47:00 1708537620000
2024/02/21 19:48:00 1708537680001
2024/02/21 19:49:00 1708537740001
2024/02/21 19:50:00 1708537800000
2024/02/21 19:51:00 1708537860001
2024/02/21 19:52:00 1708537920001
2024/02/21 19:53:00 1708537980001
2024/02/21 19:54:00 1708538040001
```

**Reply by mlusetti** [reply]
I'm astonished by your response time and response manners.
Thank you very much.
